File: swim-java/swim-core-java/swim.args/src/main/java/swim/args/Cmd.java
// Copyright 2015-2019 SWIM.AI inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package swim.args;

import swim.codec.Debug;
import swim.codec.Format;
import swim.codec.Output;
import swim.codec.Unicode;
import swim.collections.FingerTrieSeq;
import swim.util.Murmur3;

public class Cmd implements Cloneable, Debug {
  final String id;
  final String name;
  String desc;
  FingerTrieSeq<Opt> opts;
  FingerTrieSeq<Cmd> cmds;
  ExecCmd exec;
  Cmd base;

  public Cmd(String id, String name, String desc, FingerTrieSeq<Opt> opts,
             FingerTrieSeq<Cmd> cmds, ExecCmd exec, Cmd base) {
    this.id = id;
    this.name = name;
    this.desc = desc;
    this.opts = opts;
    this.cmds = cmds;
    this.exec = exec;
    this.base = base;
  }

  public Cmd base() {
    return this.base;
  }

  public String id() {
    return this.id;
  }

  public String name() {
    return this.name;
  }

  public String desc() {
    return this.desc;
  }

  public Cmd desc(String desc) {
    this.desc = desc;
    return this;
  }

  public FingerTrieSeq<Opt> opts() {
    return this.opts;
  }

  public Cmd opt(Opt opt) {
    this.opts = this.opts.appended(opt);
    return this;
  }

  public Cmd opt(String opt) {
    return opt(Opt.of(opt));
  }

  public FingerTrieSeq<Cmd> cmds() {
    return this.cmds;
  }

  public Cmd cmd(Cmd cmd) {
    this.cmds = this.cmds.appended(cmd);
    return this;
  }

  public Cmd cmd(String cmd) {
    return cmd(Cmd.of(cmd));
  }

  public Opt getOpt(String name) {
    for (int i = 0, n = this.opts.size(); i < n; i += 1) {
      final Opt opt = this.opts.get(i);
      if (name.equals(opt.name)) {
        return opt;
      }
    }
    throw new IllegalArgumentException("undefined opt: " + name);
  }

  public Cmd parse(String[] params) {
    return parse(params, 1);
  }

  public Cmd parse(String[] params, int paramIndex) {
    final int paramCount = params.length;
    final int optCount = this.opts.size();
    final int cmdCount = this.cmds.size();

    while (paramIndex < paramCount) {
      final String param = params[paramIndex];
      for (int cmdIndex = 0; cmdIndex < cmdCount; cmdIndex += 1) {
        final Cmd cmd = this.cmds.get(cmdIndex);
        if (param.equals(cmd.name)) {
          final Cmd subCmd = cmd;
          subCmd.base = this;
          return subCmd.parse(params, paramIndex + 1);
        }
      }
      paramIndex += 1;
      final int argLength = param.length();
      if (argLength > 2 && param.charAt(0) == '-' && param.charAt(1) == '-') {
        final String name = param.substring(2);
        for (int optIndex = 0; optIndex < optCount; optIndex += 1) {
          final Opt opt = this.opts.get(optIndex);
          if (name.equals(opt.name)) {
            opt.defs += 1;
            paramIndex = opt.parse(params, paramIndex);
          }
        }
      } else if (argLength > 1 && param.charAt(0) == '-') {
        for (int flagIndex = 1; flagIndex < argLength; flagIndex += 1) {
          final char flag = param.charAt(flagIndex);
          for (int optIndex = 0; optIndex < optCount; optIndex += 1) {
            final Opt opt = this.opts.get(optIndex);
            if (flag == opt.flag) {
              opt.defs += 1;
              if (argLength == 2) {
                paramIndex = opt.parse(params, paramIndex);
              }
            }
          }
        }
      }
    }
    return this;
  }

  public ExecCmd exec() {
    return this.exec;
  }

  public Cmd exec(ExecCmd exec) {
    this.exec = exec;
    return this;
  }

  public void run() {
    if (this.exec != null) {
      this.exec.exec(this);
    }
  }

  public Cmd helpCmd() {
    return this.cmd(Cmd.help());
  }

  public void writeFullName(Output<String> output) {
    if (this.base != null) {
      this.base.writeFullName(output);
      output.write(' ');
    }
    output.write(this.name);
  }

  public void writeHelp(Output<String> output) {
    output.write("Usage: ");
    this.writeFullName(output);
    final int optCount = this.opts.size();
    if (optCount != 0) {
      output.write(" [options]");
    }
    final int cmdCount = this.cmds.size();
    if (cmdCount != 0) {
      output.write(" <command>");
    }
    output.writeln();
    if (optCount != 0) {
      output.writeln();
      output.writeln("Options:");
      for (int optIndex = 0; optIndex < optCount; optIndex += 1) {
        final Opt opt = this.opts.get(optIndex);
        if (opt.flag != 0) {
          output.write("  -").write(opt.flag).write(", --").write(opt.name);
        } else {
          output.write("      --").write(opt.name);
        }
        int optLength = opt.name.length();
        final int argCount = opt.args.size();
        for (int argIndex = 0; argIndex < argCount; argIndex += 1) {
          final Arg arg = opt.args.get(argIndex);
          output.write(' ').write('<').write(arg.name).write('>');
          optLength += 2 + arg.name.length() + 1;
          if (arg.optional) {
            output.write('?');
            optLength += 1;
          }
        }
        for (int i = optLength; i < 15; i += 1) {
          output.write(' ');
        }
        if (opt.desc != null) {
          output.write(' ').write(opt.desc);
        }
        output.writeln();
      }
    }
    if (cmdCount != 0) {
      output.writeln();
      output.writeln("Commands:");
      for (int cmdIndex = 0; cmdIndex < cmdCount; cmdIndex += 1) {
        final Cmd cmd = this.cmds.get(cmdIndex);
        output.write("  ").write(cmd.name);
        for (int i = cmd.name.length(); i < 20; i += 1) {
          output.write(' ');
        }
        if (cmd.desc != null) {
          output.write("  ").write(cmd.desc);
        }
        output.writeln();
      }
    }
  }

  public String toHelp() {
    final Output<String> output = Unicode.stringOutput();
    this.writeHelp(output);
    return output.bind();
  }

  public boolean canEqual(Object other) {
    return other instanceof Cmd;
  }

  @Override
  public boolean equals(Object other) {
    if (this == other) {
      return true;
    } else if (other instanceof Cmd) {
      final Cmd that = (Cmd) other;
      return that.canEqual(this) && this.id.equals(that.id) && this.name.equals(that.name)
          && (this.desc == null ? that.desc == null : this.desc.equals(that.desc))
          && this.opts.equals(that.opts) && this.cmds.equals(that.cmds)
          && (this.exec == null ? that.exec == null : this.exec.equals(that.exec));
    }
    return false;
  }

  @Override
  public int hashCode() {
    if (hashSeed == 0) {
      hashSeed = Murmur3.seed(Cmd.class);
    }
    return Murmur3.mash(Murmur3.mix(Murmur3.mix(Murmur3.mix(Murmur3.mix(Murmur3.mix(Murmur3.mix(
        hashSeed, this.id.hashCode()), this.name.hashCode()), Murmur3.hash(this.desc)),
        opts.hashCode()), cmds.hashCode()), Murmur3.hash(this.exec)));
  }

  @Override
  public void debug(Output<?> output) {
    output = output.write("Cmd").write('.').write("of").write('(').debug(this.name).write(')');
    if (this.desc != null) {
      output = output.write('.').write("flag").write('(').debug(this.desc).write(')');
    }
    final int optCount = this.opts.size();
    for (int optIndex = 0; optIndex < optCount; optIndex += 1) {
      final Opt opt = this.opts.get(optIndex);
      output = output.write('.').write("opt").write('(').debug(opt).write(')');
    }
    final int cmdCount = this.cmds.size();
    for (int cmdIndex = 0; cmdIndex < cmdCount; cmdIndex += 1) {
      final Cmd cmd = this.cmds.get(cmdIndex);
      output = output.write('.').write("cmd").write('(').debug(cmd).write(')');
    }
    if (this.exec != null) {
      output = output.write('.').write("exec").write('(').debug(this.exec).write(')');
    }
    if (this.base != null) {
      output = output.write('.').write("base").write('(').debug(this.base).write(')');
    }
  }

  @Override
  public String toString() {
    return Format.debug(this);
  }

  @Override
  public Cmd clone() {
    final int optCount = this.opts.size();
    FingerTrieSeq<Opt> opts = FingerTrieSeq.empty();
    for (int i = 0; i < optCount; i += 1) {
      opts = opts.appended(this.opts.get(i).clone());
    }
    final int cmdCount = this.cmds.size();
    FingerTrieSeq<Cmd> cmds = FingerTrieSeq.empty();
    for (int i = 0; i < cmdCount; i += 1) {
      cmds = cmds.appended(this.cmds.get(i).clone());
    }
    return new Cmd(this.id, this.name, this.desc, opts, cmds, this.exec, this.base);
  }

  private static int hashSeed;

  public static Cmd of(String id, String name) {
    return new Cmd(id, name, null, FingerTrieSeq.empty(), FingerTrieSeq.empty(), null, null);
  }

  public static Cmd of(String id) {
    return new Cmd(id, id, null, FingerTrieSeq.empty(), FingerTrieSeq.empty(), null, null);
  }

  public static Cmd help() {
    return new Cmd("help", "help", null, FingerTrieSeq.empty(), FingerTrieSeq.empty(), new ExecHelpCmd(), null);
  }
}

final class ExecHelpCmd implements ExecCmd {
  @Override
  public void exec(Cmd cmd) {
    if (cmd.base != null) {
      System.out.println(cmd.base.toHelp());
    }
  }
}

File: swim-js/swim-runtime-js/swim-core-js/@swim/args/main/Cmd.ts
// Copyright 2015-2021 Swim Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import {Equals, Arrays} from "@swim/util";
import {Output, Debug, Format, Unicode} from "@swim/codec";
import {AnyOpt, Opt} from "./Opt";
import {AnyArg, Arg} from "./Arg";

export type ExecCmd = (this: Cmd, opts: {[name: string]: string | undefined}, args: string[]) => void;

export type AnyCmd = Cmd | CmdInit | string;

export interface CmdInit {
  id?: string;
  name: string;
  desc?: string;
  opts?: AnyOpt[];
  args?: AnyArg[];
  cmds?: AnyCmd[];
  exec?: ExecCmd | null;
}

export class Cmd implements Equals, Debug {
  constructor(id: string, name: string, desc: string | undefined, opts: Opt[],
              args: Arg[], cmds: Cmd[], exec: ExecCmd | null, base: Cmd | null) {
    Object.defineProperty(this, "id", {
      value: id,
      enumerable: true,
    });
    Object.defineProperty(this, "name", {
      value: name,
      enumerable: true,
    });
    Object.defineProperty(this, "desc", {
      value: desc,
      enumerable: true,
      configurable: true,
    });
    Object.defineProperty(this, "opts", {
      value: opts,
      enumerable: true,
      configurable: true,
    });
    Object.defineProperty(this, "args", {
      value: args,
      enumerable: true,
      configurable: true,
    });
    Object.defineProperty(this, "cmds", {
      value: cmds,
      enumerable: true,
      configurable: true,
    });
    Object.defineProperty(this, "exec", {
      value: exec,
      enumerable: true,
      configurable: true,
    });
    Object.defineProperty(this, "base", {
      value: base,
      enumerable: true,
      configurable: true,
    });
  }

  readonly id!: string;

  readonly name!: string;

  readonly desc!: string | undefined;

  withDesc(desc: string | undefined): this {
    Object.defineProperty(this, "desc", {
      value: desc,
      enumerable: true,
      configurable: true,
    });
    return this;
  }

  readonly opts!: ReadonlyArray<Opt>;

  withOpt(opt: AnyOpt): this {
    opt = Opt.fromAny(opt);
    (this.opts as Opt[]).push(opt);
    return this;
  }

  readonly args!: ReadonlyArray<Arg>;

  withArg(arg: AnyArg): this {
    arg = Arg.fromAny(arg);
    (this.args as Arg[]).push(arg as Arg);
    return this;
  }

  readonly cmds!: ReadonlyArray<Cmd>;

  withCmd(cmd: AnyCmd): this {
    cmd = Cmd.fromAny(cmd);
    (this.cmds as Cmd[]).push(cmd);
    return this;
  }

  withHelpCmd(): this {
    this.withCmd(Cmd.help());
    if (this.exec === null) {
      this.onExec(execDefaultHelpCmd);
    }
    return this;
  }

  getOpt(name: string): Opt {
    const opts = this.opts;
    for (let i = 0, n = opts.length; i < n; i += 1) {
      const opt = opts[i]!;
      if (name === opt.name) {
        return opt;
      }
    }
    throw new Error("undefined opt: " + name);
  }

  getArg(index: number = 0): Arg | null {
    const arg = this.args[index];
    return arg !== void 0 ? arg : null;
  }

  getValue(index: number = 0): string | undefined {
    const arg = this.args[index];
    return arg !== void 0 ? arg.value : void 0;
  }

  parse(params?: string[], paramIndex?: number): Cmd {
    if (params === void 0) {
      params = process.argv;
      if (paramIndex === void 0) {
        paramIndex = 2;
      }
    } else if (paramIndex === void 0) {
      paramIndex = 1;
    }
    const paramCount = params.length;
    const opts = this.opts;
    const optCount = opts.length;
    const args = this.args;
    const argCount = args.length;

    if (paramIndex < paramCount) {
      const cmds = this.cmds;
      const cmdCount = cmds.length;
      for (let cmdIndex = 0; cmdIndex < cmdCount; cmdIndex += 1) {
        const cmd = cmds[cmdIndex]!;
        if (params[paramIndex] === cmd.name) {
          cmd.withBase(this);
          return cmd.parse(params, paramIndex + 1);
        }
      }
    }

    let argIndex = 0;
    while (paramIndex < paramCount) {
      const param = params[paramIndex]!;
      paramIndex += 1;
      const argLength = param.length;
      if (argLength > 2 && param.charCodeAt(0) === 45/*'-'*/ && param.charCodeAt(1) === 45/*'-'*/) {
        const name = param.substring(2);
        for (let optIndex = 0; optIndex < optCount; optIndex += 1) {
          const opt = opts[optIndex]!;
          if (name === opt.name) {
            opt.def();
            paramIndex = opt.parse(params, paramIndex);
          }
        }
      } else if (argLength > 1 && param.charCodeAt(0) === 45/*'-'*/) {
        for (let flagIndex = 1; flagIndex < argLength; flagIndex += 1) {
          const flag = param.charAt(flagIndex);
          for (let optIndex = 0; optIndex < optCount; optIndex += 1) {
            const opt = opts[optIndex]!;
            if (flag === opt.flag) {
              opt.def();
              if (argLength === 2) {
                paramIndex = opt.parse(params, paramIndex);
              }
            }
          }
        }
      } else if (argIndex < argCount) {
        const arg = args[argIndex]!;
        arg.withValue(param);
        argIndex += 1;
      }
    }
    return this;
  }

  readonly exec!: ExecCmd | null;

  onExec(exec: ExecCmd | null): this {
    Object.defineProperty(this, "exec", {
      value: exec,
      enumerable: true,
      configurable: true,
    });
    return this;
  }

  readonly base!: Cmd | null;

  withBase(base: Cmd | null): this {
    Object.defineProperty(this, "base", {
      value: base,
      enumerable: true,
      configurable: true,
    });
    return this;
  }

  run(): void {
    if (this.exec !== null) {
      const opts: {[name: string]: string | undefined} = {};
      const args: string[] = [];
      let cmd: Cmd | null = this;
      do {
        const optCount = cmd.opts.length;
        for (let optIndex = 0; optIndex < optCount; optIndex += 1) {
          const opt = cmd.opts[optIndex]!;
          const value = opt.getValue();
          if (value !== void 0) {
            opts[opt.name] = value;
          } else if (opt.defs) {
            opts[opt.name] = void 0;
          }
        }
        const argCount = cmd.args.length;
        for (let argIndex = argCount - 1; argIndex >= 0; argIndex -= 1) {
          const arg = cmd.args[argIndex]!;
          const argValue = arg.value;
          if (argValue !== void 0) {
            args.unshift(argValue);
          }
        }
        cmd = cmd.base;
      } while (cmd !== null);
      this.exec.call(this, opts, args);
    }
  }

  /** @hidden */
  writeFullName(output: Output): void {
    if (this.base !== null) {
      this.base.writeFullName(output);
      output.write(32/*' '*/);
    }
    output.write(this.name);
  }

  /** @hidden */
  writeHelp(output: Output): void {
    output.write("Usage: ");
    this.writeFullName(output);
    const optCount = this.opts.length;
    const argCount = this.args.length;
    const cmdCount = this.cmds.length;
    if (optCount !== 0) {
      output.write(32/*' '*/).write("[options]");
    }
    if (argCount !== 0) {
      for (let argIndex = 0; argIndex < argCount; argIndex += 1) {
        const arg = this.args[argIndex]!;
        output.write(32/*' '*/).write(91/*'['*/).write(arg.name).write(93/*']'*/);
        if (arg.optional) {
          output.write(63/*'?'*/);
        }
      }
    } else if (cmdCount !== 0) {
      output.write(32/*' '*/).write("<command>");
    }
    output.writeln();
    if (optCount !== 0) {
      output.writeln();
      output.writeln("Options:");
      for (let optIndex = 0; optIndex < optCount; optIndex += 1) {
        const opt = this.opts[optIndex]!;
        if (opt.flag !== void 0) {
          output.write("  -").write(opt.flag).write(", --").write(opt.name);
        } else {
          output.write("      --").write(opt.name);
        }
        let optLength = opt.name.length;
        const optArgCount = opt.args.length;
        for (let optArgIndex = 0; optArgIndex < optArgCount; optArgIndex += 1) {
          const optArg = opt.args[optArgIndex]!;
          output.write(32/*' '*/).write(60/*'<'*/).write(optArg.name).write(62/*'>'*/);
          optLength += 2 + optArg.name.length + 1;
          if (optArg.optional) {
            output.write(63/*'?'*/);
            optLength += 1;
          }
        }
        for (let i = optLength; i < 15; i += 1) {
          output.write(32/*' '*/);
        }
        if (opt.desc !== void 0) {
          output.write(32/*' '*/).write(opt.desc);
        }
        output.writeln();
      }
    }
    if (cmdCount !== 0) {
      output.writeln();
      output.writeln("Commands:");
      for (let cmdIndex = 0; cmdIndex < cmdCount; cmdIndex += 1) {
        const cmd = this.cmds[cmdIndex]!;
        output.write("  ").write(cmd.name);
        for (let i = cmd.name.length; i < 20; i += 1) {
          output.write(32/*' '*/);
        }
        if (cmd.desc !== void 0) {
          output.write("  ").write(cmd.desc);
        }
        output.writeln();
      }
    }
  }

  toHelp(): string {
    const output = Unicode.stringOutput();
    this.writeHelp(output);
    return output.bind();
  }

  equals(that: unknown): boolean {
    if (this === that) {
      return true;
    } else if (that instanceof Cmd) {
      return this.id === that.id && this.name === that.name && this.desc === that.desc
          && Arrays.equal(this.opts, that.opts) && Arrays.equal(this.args, that.args)
          && Arrays.equal(this.cmds, that.cmds) && Equals(this.exec, that.exec)
          && Equals(this.base, that.base);
    }
    return false;
  }

  debug<T>(output: Output<T>): Output<T> {
    output = output.write("Cmd").write(46/*'.'*/).write("of").write(40/*'('*/).debug(this.name).write(41/*')'*/);
    if (this.desc !== void 0) {
      output = output.write(46/*'.'*/).write("desc").write(40/*'('*/).debug(this.desc).write(41/*')'*/);
    }
    const opts = this.opts;
    for (let optIndex = 0, optCount = opts.length; optIndex < optCount; optIndex += 1) {
      const opt = opts[optIndex]!;
      output = output.write(46/*'.'*/).write("opt").write(40/*'('*/).debug(opt).write(41/*')'*/);
    }
    const args = this.args;
    for (let argIndex = 0, argCount = args.length; argIndex < argCount; argIndex += 1) {
      const arg = args[argIndex]!;
      output = output.write(46/*'.'*/).write("arg").write(40/*'('*/).debug(arg).write(41/*')'*/);
    }
    const cmds = this.cmds;
    for (let cmdIndex = 0, cmdCount = cmds.length; cmdIndex < cmdCount; cmdIndex += 1) {
      const cmd = cmds[cmdIndex];
      output = output.write(46/*'.'*/).write("cmd").write(40/*'('*/).debug(cmd).write(41/*')'*/);
    }
    if (this.exec !== null) {
      output = output.write(46/*'.'*/).write("exec").write(40/*'('*/).debug(this.exec).write(41/*')'*/);
    }
    if (this.base !== null) {
      output = output.write(46/*'.'*/).write("base").write(40/*'('*/).debug(this.base).write(41/*')'*/);
    }
    return output;
  }

  toString(): string {
    return Format.debug(this);
  }

  clone(): Cmd {
    const optCount = this.opts.length;
    const opts = new Array<Opt>(optCount);
    for (let i = 0; i < optCount; i += 1) {
      opts[i] = this.opts[i]!.clone();
    }
    const argCount = this.args.length;
    const args = new Array<Arg>(argCount);
    for (let i = 0; i < argCount; i += 1) {
      args[i] = this.args[i]!.clone();
    }
    const cmdCount = this.cmds.length;
    const cmds = new Array<Cmd>(cmdCount);
    for (let i = 0; i < cmdCount; i += 1) {
      cmds[i] = this.cmds[i]!.clone();
    }
    return new Cmd(this.id, this.name, this.desc, opts, args, cmds, this.exec, this.base);
  }

  static create(id: string, name?: string, desc?: string, anyOpts?: AnyOpt[],
                anyArgs?: AnyArg[], anyCmds?: AnyCmd[], exec: ExecCmd | null = null): Cmd {
    if (name === void 0) {
      name = id;
    }
    const optCount = anyOpts !== void 0 ? anyOpts.length : 0;
    const opts = new Array<Opt>(optCount);
    for (let optIndex = 0; optIndex < optCount; optIndex += 1) {
      opts[optIndex] = Opt.fromAny(anyOpts![optIndex]!);
    }
    const argCount = anyArgs !== void 0 ? anyArgs.length : 0;
    const args = new Array<Arg>(argCount);
    for (let argIndex = 0; argIndex < argCount; argIndex += 1) {
      args[argIndex] = Arg.fromAny(anyArgs![argIndex]!);
    }
    const cmdCount = anyCmds !== void 0 ? anyCmds.length : 0;
    const cmds = new Array<Cmd>(cmdCount);
    for (let cmdIndex = 0; cmdIndex < cmdCount; cmdIndex += 1) {
      cmds[cmdIndex] = Cmd.fromAny(anyCmds![cmdIndex]!);
    }
    return new Cmd(id, name, desc, opts, args, cmds, exec, null);
  }

  static fromInit(init: CmdInit): Cmd {
    const id = init.id !== void 0 ? init.id : init.name;
    const optCount = init.opts !== void 0 ? init.opts.length : 0;
    const opts = new Array<Opt>(optCount);
    for (let optIndex = 0; optIndex < optCount; optIndex += 1) {
      opts[optIndex] = Opt.fromAny(init.opts![optIndex]!);
    }
    const argCount = init.args !== void 0 ? init.args.length : 0;
    const args = new Array<Arg>(argCount);
    for (let argIndex = 0; argIndex < argCount; argIndex += 1) {
      args[argIndex] = Arg.fromAny(init.args![argIndex]!);
    }
    const cmdCount = init.cmds !== void 0 ? init.cmds.length : 0;
    const cmds = new Array<Cmd>(cmdCount);
    for (let initIndex = 0; initIndex < cmdCount; initIndex += 1) {
      cmds[initIndex] = Cmd.fromAny(init.cmds![initIndex]!);
    }
    const exec = init.exec !== void 0 ? init.exec : null;
    return new Cmd(id, init.name, init.desc, opts, args, cmds, exec, null);
  }

  static fromAny(value: AnyCmd): Cmd {
    if (value instanceof Cmd) {
      return value;
    } else if (typeof value === "string") {
      return new Cmd(value, value, void 0, [], [], [], null, null);
    } else if (typeof value === "object" && value !== null) {
      return Cmd.fromInit(value);
    } else {
      throw new TypeError("" + value);
    }
  }

  static help(): Cmd {
    return Cmd.create("help").onExec(execHelpCmd);
  }
}

function execHelpCmd(this: Cmd, args: {[name: string]: string | undefined}) {
  if (this.base !== null) {
    console.log(this.base.toHelp());
  }
}

function execDefaultHelpCmd(this: Cmd, args: {[name: string]: string | undefined}) {
  console.log(this.toHelp());
}
